### PR TITLE
Feature/edit file locally restart sync

### LINKED ISSUE #5110: Open file for local editing should not wait till the currently-running sync completes
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you want to have the same feature implemented.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Feature description
Right now, if we have a long sync already running, or, if we have a lot of files to sync (many data on the server) and we click the "Edit locally" link in the WebUI, we will have to wait indefinitely until the sync completes before the file will get opened locally or even worse it may not be selected for sync and we will still have to wait till the sync completes before seeing the error message.

We need to abort the currently running sync (maybe make sure we are not aborting hydration of a file via VFS as this might bring troubles), and then, force-sync just one file that we want to edit locally or the folder it resides in. This will make opening a file for local editing faster.

NOTE: This will need some nice design decisions.

### PR DESCRIPTION
**Pull request by allexzander**
This will fix #5110

Changes:
- edit locally task is now put to the top priority and when it gets triggered, all other running syncs get aborted
- it is now possible to force sync only a specific file even if it was not yet synced but is not blacklisted

### CODE REVIEW

**Comment by tobiasKaminsky** [comment]
@allexzander have you closed this with purpose?

**Review by claucambra — CHANGES_REQUESTED** [review]

**Comment by allexzander** [comment]
> @allexzander have you closed this with purpose?

yes, our CI has hung, so, I've closed and reopened it just to make sure it gets through all the checks
should I not do that in future?

File: src/gui/editlocallyjob.cpp [hunk 1]
```diff
@@ -29,12 +29,12 @@ namespace OCC {
 Q_LOGGING_CATEGORY(lcEditLocallyJob, "nextcloud.gui.editlocallyjob", QtInfoMsg)
 
 EditLocallyJob::EditLocallyJob(const QString &userId,
-                                       const QString &relPath,
+                                       const QString &fullRelPath,
```
Comment (claucambra) [inline src/gui/editlocallyjob.cpp:32]: I should have clarified -- it's not clear what a "full relative path" is without digging through how this comes to be created. IMO it should have a more descriptive name or a comment in the header explaining what this parameter is supposed to be, as without context, a "full relative path" can seem to be a bit paradoxical

File: src/gui/editlocallyjob.h [hunk 2]
```diff
@@ -17,6 +17,7 @@
 #include <QObject>
 
 #include "accountstate.h"
+#include <syncfileitem.h>
```
Comment (claucambra) [inline src/gui/editlocallyjob.h:20]:
```suggestion
#include "syncfileitem.h"
```

File: src/libsync/syncfileitem.cpp [hunk 3]
```diff
@@ -13,8 +13,10 @@
  */
 
 #include "syncfileitem.h"
+#include <common/checksums.h>
```
Comment (claucambra) [inline src/libsync/syncfileitem.cpp:16]:
```suggestion
#include "common/checksums.h"
```

File: src/gui/editlocallyjob.h [hunk 4]
```diff
@@ -50,6 +51,14 @@ public slots:
...
     void startEditLocally();
 
+private:
+    void fetchRemoteFileParentInfo();
+    bool checkIfFileParentSyncIsNeeded(); // returns true if sync will be needed, false otherwise
+    void startSyncBeforeOpening();
+    void eraseBlacklistRecordForItem();
+    [[nodiscard]] const QString getRelativePathToRemoteRootForFile() const;
+    [[nodiscard]] const QString getRelativePathParent() const;
+
```
Comment (claucambra) [inline src/gui/editlocallyjob.h:61]: I would combine this private section with the private section further down the class declaration, splitting these up makes the header more difficult to read

File: src/gui/editlocallyjob.h [hunk 5]
```diff
@@ -50,6 +51,14 @@ public slots:
     void startSetup();
     void startEditLocally();
 
+private:
+    void fetchRemoteFileParentInfo();
+    bool checkIfFileParentSyncIsNeeded(); // returns true if sync will be needed, false otherwise
+    void startSyncBeforeOpening();
+    void eraseBlacklistRecordForItem();
```
Comment (claucambra) [inline src/gui/editlocallyjob.h:58]: These void methods could be private slots

File: src/gui/editlocallyjob.cpp [hunk 6]
```diff
@@ -118,42 +118,65 @@ void EditLocallyJob::remoteTokenCheckResultReceived(const int statusCode)
...
+        showError(tr("Could not find a file for local editing. Make sure it is not excluded via selective sync."), _relPath);
+        return;
+    }
+
+    if (_relPathParent == QStringLiteral("/")) {
+        proceedWithSetup();
+        return;
+    }
+
+    fetchRemoteFileParentInfo();
```
Comment (claucambra) [inline src/gui/editlocallyjob.cpp:148]: I can see the use of the new `getRelativePathParent` 🙂 I still don't think we should have this much code unrelated to the token check result processing in `remoteTokenCheckResultReceived`, though, this is slightly misleading...

I would either move all of this to a new method that then itself calls `proceedWithSetup`, or move this to `proceedWithSetup` itself

Comment (claucambra) [reply src/gui/editlocallyjob.cpp:148]: I suspect the former solution will be easier than the latter due to the job that needs to be run in `fetchRemoteFileParentInfo`

Comment (claucambra) [reply src/gui/editlocallyjob.cpp:148]: FWIW I also don't think I designed this very well when I added a call to `proceedWithSetup` in `remoteTokenCheckResultReceived`, so if there is a better design to be put here then even better

**Review by claucambra — CHANGES_REQUESTED** [review]

File: src/gui/editlocallyjob.cpp (hunk 6 above)
Comment (allexzander) [reply src/gui/editlocallyjob.cpp:148]: @claucambra Besides having a state-machine, I don't see another solution, so, let's keep that `proceedWithSetup` and I am gonna move the unrelated part out of `remoteTokenCheckResultReceived`

**Review by claucambra — APPROVED** [review]
Other than small nitpick, looks good and works great 🚀

File: src/gui/editlocallyjob.cpp [hunk 7]
```diff
@@ -118,52 +118,199 @@ void EditLocallyJob::remoteTokenCheckResultReceived(const int statusCode)
...
 }
 
+void EditLocallyJob::fetchRemoteFileParentInfo()
+{
+    Q_ASSERT(_relPathParent != QStringLiteral("/"));
+
+    if (_relPathParent == QStringLiteral("/")) {
+        qCWarning(lcEditLocallyJob) << "LsColJob must only be used for nested folders.";
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (allexzander) [reply src/gui/editlocallyjob.cpp:202]: @claucambra The assert does nothing in release mode, so, keeping it.

**Pull request merged**